src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2653,268 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
 	return false;
 }
+
+
+/*
+ * Identifiers such as schema, table, column comes from various
+ * sources(e.g. wal2json, test_decoding and source catalog) and some of them
+ * already escapes identifiers and few don't.
+ * We need to check if the identifier is already quoted or not before
+ * escaping it.
+ * Whatever we are here is not a fool proof escaping mechanism, but a best
+ * effort to make sure that the identifiers are normalized by quoting them
+ * if it is not already quoted.
+ *
+ * Here is an example:
+ * foo -> "foo"
+ * "foo" -> "foo"
+ * foo"bar -> "foo"bar"
+ * "foo -> ""foo"
+ *
+ * The goal of this normalization is to make sure that the identifiers are
+ * comparable in the context of Hash Table.
+ */
+#define NORMALIZE_IDENTIFIER(buf, bufsize, offset, ident) \
+	{ \
+		int len = strlen(ident); \
+		if (ident[0] == '"' && ident[len - 1] == '"') \
+		{ \
+			offset += sformat(buf, bufsize, "%s", ident); \
+		} \
+		else \
+		{ \
+			offset += sformat(buf, bufsize, "\"%s\"", ident); \
+		} \
+	}
+
+/*
+ * createFQColumnName creates a fully qualified column name in the form
+ * schema.table.column and normalizes the identifiers by quoting them if
+ * necessary.
+ *
+ */
+static void
+createFQColumnName(FQColumnName *qColumnName,
+				   const char *schema,
+				   const char *table,
+				   const char *column)
+{
+	int offset = 0;
+	char *buf = (char *) qColumnName;
+	const int bufsize = sizeof(FQColumnName);
+
+	NORMALIZE_IDENTIFIER(buf + offset, bufsize - offset, offset, schema);
+	offset += sformat(buf + offset, bufsize - offset, ".");
+
+	NORMALIZE_IDENTIFIER(buf + offset, bufsize - offset, offset, table);
+	offset += sformat(buf + offset, bufsize - offset, ".");
+
+	NORMALIZE_IDENTIFIER(buf + offset, bufsize - offset, offset, column);
+}
+
+
+/*
+ * isGeneratedColumn checks whether the given "column" in "schema.table" is
+ * generated or not using the cache.
+ *
+ * Returns true if the column is generated, false otherwise.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
+ */
+static bool
+isGeneratedColumn(GeneratedColumnsCache *cache,
+				  const char *schema,
+				  const char *table,
+				  const char *column)
+{
+	/*
+	 * NULL cache means that we don't have any generated columns in the
+	 * catalog.
+	 */
+	if (cache == NULL)
+	{
+		return false;
+	}
+
+	GeneratedColumnsCache *item = NULL;
+	FQColumnName qColumnName = { 0 };
+
+	(void) createFQColumnName(&qColumnName,
+							  schema, table, column);
+
+	HASH_FIND_STR(cache, qColumnName, item);
+
+	bool isGenerated = (item != NULL);
+
+	if (isGenerated)
+	{
+		log_debug("IsColumnGenerated: %s is a generated column",
+				  qColumnName);
+	}
+
+	return isGenerated;
+}
+
+
+/*
+ * populateGeneratedColumnCache is a callback function that populates the
+ * generated columns cache from the catalog.
+ */
+static bool
+populateGeneratedColumnCache(void *ctx, GeneratedColumn *column)
```

Comment:
Nitpicking: all the callback function names ends the same way, with `_hook` as a postfix. Please consider doing the same here.

Reply:
Done.